transactions: add big comment about offers and the order book.

graydon opened the pull request:
This adds a big explanatory comment about how the order book works and what the general ideas behind the limit and exact offers in the system are all about. I wrote it because the order book code is pretty opaque to me still, and I really couldn't make heads or tail of CAP-0006 without this level of pedantic self-explanation. Figured it might be useful to others.

I would _love_ a thorough read-through and correction of it if possible from @jonjove and/or @MonsieurNicolas because I'm only very very moderately convinced I got it right. I think the commentary hangs together in the sense that I think it's painting a logically-consistent explanation for things, but I am much less sure that it's _correct_ with respect to the code.

Apologies for the level of repetitiveness / redundancy of phrasing / pictures / etc. I know it's probably overkill for people familiar with the subsystem and ideas, but for me I really get completely lost every time I dig into it. Wanted a very explicit guide.

MonsieurNicolas commented:
I'll take a look at this a bit later. Glancing at it, most should probably go in the "distributed exchange" docs https://github.com/stellar/docs/blob/master/guides/concepts/exchange.md as *any* user (not just stellar-core devs) should understand it.

You can leave the technical bits about where in the code things live in comments.

vogel commented on src/transactions/OfferExchange.h:
```diff
@@ -8,6 +8,228 @@
+//
+// Orders/Offers (see terminology note above) submitted to the system are
+// actually _limit_ offers, meaning that a new order will match against any
+// stored offer in he book offering the seller a _higher_ price for their merch
```
typo: in the book

graydon replied:
fixed.

MonsieurNicolas commented on src/transactions/OfferExchange.h:
```diff
@@ -8,6 +8,228 @@
+//
+// What price is the sale made at? The price of the offer stored in the book. In
+// fact, offers _in the book_ are not limit offers, they are _exact_ offers that
+// strictly cross _at the price written on them_, never anything else. In the
```
"never anything else" is actually not true, cap0003 describes this more, but rounding causes bias that can cause a better price (when the standing offer wins) or worst price (when the standing offer is taken entirely), it's also possible that the order book decides to just throw away parts of the order to avoid violating threshold (to avoid making more than 1% error on the price)

graydon replied:
I think you mean CAP-004. Noted and added expanded section discussing.

oleganza commented:
this patch is very illuminating, thanks.

graydon commented:
I split the first half (orientation) off to https://github.com/stellar/docs/pull/453 against the docs repo, and fixed the associated suggestions here. Anything else?

MonsieurNicolas commented:
just noticed that you have a libmedida change in there, probably not what you wanted to do

graydon commented:
@MonsieurNicolas whoops! nice catch thanks, reverted the medida change.
